New Resource: aws_wafregional_geo_match_set

radeksimko opened the pull request:
## Test results

```
TF_ACC=1 go test ./aws -v -run=TestAccAWSWafRegionalGeoMatchSet_ -timeout 120m
=== RUN   TestAccAWSWafRegionalGeoMatchSet_basic
--- PASS: TestAccAWSWafRegionalGeoMatchSet_basic (40.38s)
=== RUN   TestAccAWSWafRegionalGeoMatchSet_changeNameForceNew
--- PASS: TestAccAWSWafRegionalGeoMatchSet_changeNameForceNew (65.81s)
=== RUN   TestAccAWSWafRegionalGeoMatchSet_disappears
--- PASS: TestAccAWSWafRegionalGeoMatchSet_disappears (30.67s)
=== RUN   TestAccAWSWafRegionalGeoMatchSet_changeConstraints
--- PASS: TestAccAWSWafRegionalGeoMatchSet_changeConstraints (67.57s)
=== RUN   TestAccAWSWafRegionalGeoMatchSet_noConstraints
--- PASS: TestAccAWSWafRegionalGeoMatchSet_noConstraints (31.64s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	236.116s
```

bflad commented on aws/resource_aws_wafregional_geo_match_set.go:
```diff
@@ -0,0 +1,158 @@
+		Delete: resourceAwsWafRegionalGeoMatchSetDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
```
Minor nitpick: Extraneous `&schema.Schema` here and below

bflad commented on aws/resource_aws_wafregional_geo_match_set.go:
```diff
@@ -0,0 +1,158 @@
+
+	resp, err := conn.GetGeoMatchSet(params)
+	if err != nil {
+		if isAWSErr(err, "WAFNonexistentItemException", "") {
```
Similar note here as other PR about upstream SDK issue.

Also, minor duplicate `WAF WAF` in the error below. (Aside: I kind of wonder if we should have some sort of `FriendlyName` attribute on resources)

radeksimko replied:
> I kind of wonder if we should have some sort of FriendlyName attribute on resources

It sounds like the time is coming to reopen discussion around https://github.com/hashicorp/terraform/pull/13582 😄

bflad commented on aws/resource_aws_wafregional_geo_match_set_test.go:
```diff
@@ -0,0 +1,325 @@
+			{
+				Config: testAccAWSWafRegionalGeoMatchSetConfigChangeName(geoMatchSetNewName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSWafRegionalGeoMatchSetExists("aws_wafregional_geo_match_set.test", &after),
```
I'd expect an acceptance test here that compares `&before` to `&after` (via ID or however necessary)

appilon commented on aws/resource_aws_wafregional_geo_match_set.go:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	d.Set("name", resp.GeoMatchSet.Name)
+	d.Set("geo_match_constraint", flattenWafGeoMatchConstraint(resp.GeoMatchSet.GeoMatchConstraints))
```
consider moving flattening func to `waf_helpers.go`

appilon commented on aws/resource_aws_wafregional_geo_match_set.go:
```diff
@@ -0,0 +1,158 @@
+		req := &waf.UpdateGeoMatchSetInput{
+			ChangeToken:   token,
+			GeoMatchSetId: aws.String(id),
+			Updates:       diffWafGeoMatchSetConstraints(oldConstraints, newConstraints),
```
consider moving diff func to `waf_helpers.go`

radeksimko commented:
Addressed all your feedback and reran acceptance tests:

```
=== RUN   TestAccAWSWafRegionalGeoMatchSet_basic
--- PASS: TestAccAWSWafRegionalGeoMatchSet_basic (37.84s)
=== RUN   TestAccAWSWafRegionalGeoMatchSet_changeNameForceNew
--- PASS: TestAccAWSWafRegionalGeoMatchSet_changeNameForceNew (105.75s)
=== RUN   TestAccAWSWafRegionalGeoMatchSet_disappears
--- PASS: TestAccAWSWafRegionalGeoMatchSet_disappears (30.65s)
=== RUN   TestAccAWSWafRegionalGeoMatchSet_changeConstraints
--- PASS: TestAccAWSWafRegionalGeoMatchSet_changeConstraints (68.88s)
=== RUN   TestAccAWSWafRegionalGeoMatchSet_noConstraints
--- PASS: TestAccAWSWafRegionalGeoMatchSet_noConstraints (34.05s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	277.202s
```

bflad commented:
This has been released in [version 1.13.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1130-march-28-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
